feat: Application dependencies (#15280)
argoproj/argo-cd: docs/user-guide/app_dependencies.md

# Application dependencies

Starting from Argo CD v2.9, you can natively model dependencies between Applications. This new mechanism is intended to replace the prior way of modeling dependencies when using the app-of-apps pattern with sync waves on the Application custom resources.

Using the new dependency mechanism, users are not bound to the app-of-apps pattern anymore and can define dependencies between applications using application selectors to define a directed, acyclic graph for dependency modeling. This means that an Application which defines one or more dependencies do not only depend on its direct dependencies, but to the dependencies of its dependencies as well.

Application dependencies support both, manual and automatic sync policies.

## Feature state and caveats

This feature is in [beta state](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md#beta). It will be polished across the upcoming releases.

Bug reports and enhancement requests are most welcome.

## Limitations and known issues

A list of things that are good to know before getting started with Application dependencies:

> * An Application can only depend on other Applications that use the same AppProject and are in the same namespace as the dependent Application.

Reviewer: Are there any plans to remove the same namespace restriction in the future? I can understand that the dependent applications should be under the same AppProject, but regarding the namespaces they may be spread over the namespaces, especially in the cluster administrator point of view.

Reviewer: nvm, I completely misunderstood the Application namespace and the target namespace that is defined in Application. There may be some use cases because we already have the concept of [Applications in any namespace](https://argo-cd.readthedocs.io/en/stable/operator-manual/app-any-namespace/), but I don't have a requirement about it.

Reviewer 2: I second the part about depending on an `Application` belonging to another `AppProject`. As a specific example we have a bunch of "baseapps" which are managed by the DevOps team, which belongs to a baseapps project. These applications, from a developer's perspective are read-only (they can view these apps, but not do anything apart from that). 

It would be useful to have a dependency between an `Application` belonging to the baseapp project, and an `Application` which belongs to another `AppProject` (in our case, every team has their own dedicated `AppProject`).

A contrived example here (this specific example would be better handled by retrying the dependent application, but just to illustrate the dependency):

1. `aws-load-balancer-controller` belongs to the baseapp project
2. `foo-app` belonging to AppProject A needs `aws-load-balancer-controller` to be up before starting
3. `bar-app` belonging to AppProject B  also needs `aws-load-balancer-controller` to be up before starting

Author: I feel this is a valid requirement, and I had initially implemented it, but we need to be careful here. Because in a multi-tenant system, you may not want someone with access in project A to even know about project B, let alone the Applications in that project or their sync/health status.

@crenshaw-dev and I also had a brief discussion about this today, but without a final conclusion. 

How about having that as a setting in argocd-cm (or as a command line switch to the application controller, or somewhere else) that allows cross-project and cross-namespace dependencies and that must be explicitly turned on by the Argo CD administrator?

Reviewer 3: I presume that there's a default AppProject? (It isn't clear from this text, and I'd hate to have to, as an end user, figure out if I have to opt into another feature first before I could use this one.)

Author: @jsoref  AppProjects are one of the core concepts of Argo CD, so I feel users should know about it already

Reviewer 3: I have no idea. And I've been using it for years. (Seriously)

Reviewer 2: > How about having that as a setting in argocd-cm (or as a command line switch to the application controller, or somewhere else) that allows cross-project and cross-namespace dependencies and that must be explicitly turned on by the Argo CD administrator?

It makes sense for this to be opt-in by the Argo CD admin. Ideally this would be a configuration in the AppProject itself, something like

```yaml
# This would allow appprojects to depend on the app project declaring the stanza below
permitDependsOnFor:
- appProjectA
- appProjectB
```

I'm not sure how this would interplay with self-serve app projects, so if there's something preventing this from being possible then I guess we would need to configure this directly in the application controller. Or perhaps another CRD which can be referenced by AppProjects which in turn would only be allowed to be managed by Argo CD admins?

Author: Sorry for the late reply here, @blakepettersson 

>I'm not sure how this would interplay with self-serve app projects

I don't think self-serve app projects are a good idea, and I think we don't have to consider this scenario. So your example makes sense. I think we might want to bring this enhancement in the next iteration of this feature, if it's something our users want.

* A circular dependency defined anywhere in the dependency chain is considered to be an error and will prevent any application in the chain from syncing.

> * Syncing an application that is dependent on other apps will not trigger a sync of the dependencies. However, manual sync of a dependent application whose dependencies are not satisfied will block.

Reviewer 4: What is the definition of a satisfied dependency vs not satisfied. I think we need to clarify it in the doc. 
If service A depends on service B and B is in degraded state, what will happen when trying to sync A? What are all conditions that would block Service A to sync?

Author: Agree, fair call. In short, the conditions are the same as for sync waves (i.e. application's sync status must be `Synced` and health status be `Healthy`). I'll make sure to write that out in the docs.

Author: I updated the docs. If it still doesn't make sense what I wrote (I have the tendency to assume too much implicit knowledge by the user), please feel free to call that out :)

Reviewer 4: Thank you. Added a small suggestion but it looks good now!

* Dependencies currently can only be configured in a declarative manner; the UI and CLI do not yet have first class support for configuring dependencies.

> 

Reviewer 2: Is there a way to also extend this so that `Application`s can be deleted in reverse order? IMO it's not super interesting to be able to depend on another application, but on the other hand it becomes super important to delete applications in reverse order (e.g for the example above, if `aws-load-balancer-controller` gets deleted before the dependent apps, we'd have orphaned load balancers).

Ideally there'd be a way to set the dependency policy, something like

1. `both` (use the DAG both when syncing an application and deleting it)
2. `delete-only` 
3. `sync-only`

Reviewer 5: IMO, we should block a deletion of an App in the first iteration itself if there are any dependent applications on the current application. 

The deletion of an Application in the dependency chain could lead to sync issues in the applications depending on the deleted application. For example, consider the dependency chain, A -> B -> C -> D (A depends on B, B depends on C, etc.). If the application B is deleted and the `blockOnEmpty` is not set to true, this will break the Application A getting synced.

WDYT?

Author: While I definitely see the issue behind of what you're saying, I'm not sure I agree with the criticality. 

I thought a little bit more about the deletion subject recently, and I think we might not want do bring it as part of the MVP. My concerns basically revolve around a couple of thoughts, that potentially need further discussions, investigations and validations:

* Deletion in Kubernetes can only be blocked by setting a finalizer on a resource or by modeling owner references with `blockOwnerDeletion` set to true. Owner references only work within the same namespace, so I think they're out of the picture for now. Leaves us with finalizers, and as long as a finalizer is set, the Kubernetes GC will not reap a given resource. So we'd need to put a new finalizer on every Application that is a dependency of another Application. However, due to the dynamic nature of this finalizer, it's not clear what that means in term of sync status for the Application (for example, user has put a resources finalizer manually to the application, and we inject a dependency finalizer).

* Once a deletion is triggered, it cannot be easily aborted (if at all). For example, if a user deletes an Application that is a dependency of something else, the Application will stay in terminating mode as long as the dependent still exist. The only way to forcefully delete that Application is to remove the finalizer from the resource, but right now, there is no way I'm aware of to do this via the UI or CLI. That might come unexpected for many, especially when this finalizer is introduced to the Application on behalf of another Application (the dependent), and not in direct control of the Application's owner.

Not saying this is impossible or we shouldn't do it, but IMO it's out of scope for the MVP. The problem we are  trying to solve first is to allow ordered sync of Applications without using the app-of-apps pattern and sync waves.

The deletion topic makes a good follow up enhancement from my point of view.

Reviewer 5: It makes sense to take care of it in the next iteration. But in that case, lets explicitly call it out in the docs that the deletion of any application from the dependency chain can break the application sync and the users need to be careful before deleting the application.

Author: I called it out in the docs.

      - other-app
```

The above YAML will have the Application depend on the Applications `some-app` and `other-app`. 

The `namePattern` matcher supports shell-style glob matching, so the following example would match all Applications whose name is starting with `stage1-`:

```yaml
spec:
  dependsOn:
    selectors:
    - namePattern:
      - stage1-*
```

Regular expressions are not supported.

## Configuring behaviour of dependency resolution

There are several options to configure the behaviour of dependency resolution. Please note that these settings are not inherited to dependencies, but must be configured for each dependant uniquely.
### Delayed start of automated sync

In certain scenarios, you may want to delay the automatic sync process when there are dependencies configured between applications that are created at the same time (i.e. synced through an app of apps pattern from Git). This may allow the synchronization of all applications to succeed before the dependency tree is resolved and the sync is actually started in the right order.

To configure the sync delay, specify the number of seconds to delay in the field `syncDelay` as follows:

```yaml
spec:
  dependsOn:

>     syncDelay: 5

Reviewer 6: Is delay the only options when is a dependency considered meet, is the requirement just synced what about health checks or  status checks within the resources. I am just thinking out loud an example being I have an operator that creates AWS resources the syncing of that resource does not mean that it's healthy/ready but that my ALB CRD has some health check or `.status.alb.dnsName` being set. This could maybe be documented somewhere?

Reviewer 7: It waits for dependent applications to become healthy. So you may want to create a [custom health check](https://argo-cd.readthedocs.io/en/stable/operator-manual/health/#custom-health-checks) for your CRD?

Author: Yes, the dependencies feature behave like sync waves, i.e. it inspects the dependencies' sync and health states. 

The `syncDelay` controls how long to wait before checking the status. For example, consider all applications created at the same time, with auto sync turned on. The application controller will start reconciling & syncing whichever app wins the race, so if a dependent app wins the race to a dependency, the dependency's health status might not yet reflect a progressing state. So `syncDelay` gives the application controller a chance to properly inspect dependencies' status. This is also similar to how [sync waves behave](https://github.com/argoproj/argo-cd/blob/129cf5370f9e2c6f99c9a5515099250a7ba42099/controller/sync.go#L534).

Author: FTR, I updated the docs to state how dependencies are satisfied

# Application dependencies

> Starting from Argo CD v2.9, you can natively model dependencies between Applications. This new mechanism is intended to replace the prior way of modeling dependencies when using the app-of-apps pattern with sync waves on the Application custom resources.

Reviewer 3: I'm not a huge fan of burning version numbers like this (2.9) into commit contents. They make predictions that often fail.

I'd _somewhat_ favor "if you're reading this, and using the version related to this documentation" or something like that. Or even a placeholder macro and letting the doc build system fill it in, and then at release branch point burn the versions into the main branch documentation.

Author: ```suggestion
Argo CD lets you natively model dependencies between Applications. This mechanism is intended to replace the prior way of modeling dependencies when using the app-of-apps pattern with sync waves on the Application custom resources.
```

Author: I agree. Does above suggestion sound better?

> Using the new dependency mechanism, users are not bound to the app-of-apps pattern anymore and can define dependencies between applications using application selectors to define a directed, acyclic graph for dependency modeling. This means that an Application which defines one or more dependencies do not only depend on its direct dependencies, but to the dependencies of its dependencies as well.

Reviewer 3: `the new` will eventually be wrong.

https://prague-info.info/old_new_synagogue/

Author: ```suggestion
Using the native dependency mechanism, users are not bound to the app-of-apps pattern anymore and can define dependencies between applications using application selectors to define a directed, acyclic graph for dependency modeling. This means that an Application which defines one or more dependencies do not only depend on its direct dependencies, but to the dependencies of its dependencies as well.
```

Author: How about this suggestion?

> Using the new dependency mechanism, users are not bound to the app-of-apps pattern anymore and can define dependencies between applications using application selectors to define a directed, acyclic graph for dependency modeling. This means that an Application which defines one or more dependencies do not only depend on its direct dependencies, but to the dependencies of its dependencies as well.

Reviewer 3: ```suggestion
Using the new dependency mechanism, users are not bound to the app-of-apps pattern anymore and can define dependencies between applications using application selectors to define a directed, acyclic graph for dependency modeling. This means that an Application which defines one or more dependencies is not limited to only depending on its direct dependencies, but can also require the dependencies of its dependencies as well.
```

Author: This sounds to me like taking dependencies of dependencies into account would be optional, though, but it isn't optional.

Reviewer 3: ```suggestion
Using the native dependency mechanism, can define dependencies between applications using application selectors to define a directed, acyclic graph for dependency modeling. This means that an Application which defines one or more dependencies can reference both direct dependencies and the dependencies of its dependencies.

This replaces the previous app-of-apps pattern where dependencies were limited to direct dependencies.
```

Application dependencies support both, manual and automatic sync policies.

## Sync behaviour

Defining dependencies for an Application will affect that Application's sync behaviour. When a sync is started, Argo CD will evaluate the sync and health status of all other Applications in the dependency tree. Only if all of the other Applications that the Application is dependent upon have a sync status of `Synced` and a health status of `Healthy`, the sync will proceed. Until all conditions are met, the sync will stay in a progressing state and a message will indicate that the sync is waiting for dependencies. This behaviour is similar to sync waves, where a sync progresses only to the next wave when the current wave's resources are all synced and healthy.

> As a fallback mechanism, when using manual sync with the force option, Argo CD will not perform dependency resolution and the sync will proceed regardless of the Application's dependencies' states.

Reviewer 3: ```suggestion
As a fallback mechanism, when using manual sync with the force option, Argo CD will not perform dependency resolution and the sync will proceed regardless of the states of the Application's dependencies.
```

## Feature state and caveats

This feature is in [beta state](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md#beta). It will be polished across the upcoming releases.

Bug reports and enhancement requests are most welcome.

## Limitations and known issues

A list of things that are good to know before getting started with Application dependencies:

* An Application can only depend on other Applications that use the same AppProject and are in the same namespace as the dependent Application.

> * A circular dependency defined anywhere in the dependency chain is considered to be an error and will prevent any application in the chain from syncing.

Author: > Does it mean we will find a cyclic dependency and surface the cyclic dependency as an error?

Yes, a cyclic dependency surfaces as an error condition for the Application

> Would force sync ignore cyclic dependencies as well?

No, a force sync only lets the user override unsatisfied dependencies, it doesn't let you override any error condition set on the Application.

> * A circular dependency defined anywhere in the dependency chain is considered to be an error and will prevent any application in the chain from syncing.

Reviewer 3: ```suggestion
* A circular dependency defined anywhere in the dependency chain is considered to be an error and will prevent all applications in the circular dependency loop from syncing.
```

> * Syncing an application that is dependent on other apps will not trigger a sync of the dependencies. However, manual sync of a dependent application whose dependencies are not satisfied will block unless force sync is used.

Author: Yes, makes sense. I'll update the docs.

> * Syncing an application that is dependent on other apps will not trigger a sync of the dependencies. However, manual sync of a dependent application whose dependencies are not satisfied will block unless force sync is used.

Author: Put the warning in the docs.

Each selector can match Applications according to one or more properties. If you specify more than one matcher in a selector, each matcher has to match in order for an Application to be considered a dependency. For example, if you specify a `labelSelector` and a `namePattern`, both will have to match in order for an Application to be selected.

Additionally, the behaviour of dependency resolution can be configured on a per Application basis.

## Dependency selectors

Currently, the available properties to match against are `labels` and the `name` of the Application. Each selector can yield zero or more applications.

### Matching on labels

It is possible and recommended to use a standard label selector to select the dependencies for your Application:

```yaml
spec:
  dependsOn:
    selectors:
    - labelSelector:
        matchLabels:
            name: some-app
```

The above YAML will have the Application depend on any other Application that matches the given label selector. In this case, any application carrying a label `name` with the value of `some-app`.

### Matching on name patterns

> Dependencies can be matched on the name of Applications by specifying the `namePattern` selector. The `namePattern` selector takes a list of Application name patterns to match against the Applications' `.metadata.name` property. Specifying multiple patterns 

Reviewer 3: ```suggestion
Dependencies can be matched on the name of Applications by specifying the `namePattern` selector. The `namePattern` selector takes a list of Application name patterns to match against the Applications' `.metadata.name` property. Specifying multiple patterns:
```

It feels like there are words / whitespace missing near `Specifying multiple patterns`

```yaml
spec:
  dependsOn:
    selectors:
    - namePattern:
      - some-app
      - other-app
```

The above YAML will have the Application depend on the Applications `some-app` and `other-app`. 

The `namePattern` matcher supports shell-style glob matching, so the following example would match all Applications whose name is starting with `stage1-`:

```yaml
spec:
  dependsOn:
    selectors:
    - namePattern:
      - stage1-*
```

Regular expressions are not supported.

## Configuring behaviour of dependency resolution

> There are several options to configure the behaviour of dependency resolution. Please note that these settings are not inherited to dependencies, but must be configured for each dependant uniquely.

Reviewer 3: ```suggestion
There are several options to configure the behaviour of dependency resolution. Please note that these settings are not inherited to dependencies, but must be configured for each dependent uniquely.

```

### Delayed start of automated sync

In certain scenarios, you may want to delay the automatic sync process when there are dependencies configured between applications that are created at the same time (i.e. synced through an app of apps pattern from Git). This may allow the synchronization of all applications to succeed before the dependency tree is resolved and the sync is actually started in the right order.

To configure the sync delay, specify the number of seconds to delay in the field `syncDelay` as follows:

```yaml
spec:
  dependsOn:
    syncDelay: 5
```

> Please note that any delay you configure here will delay _any_ automatic sync, regardless of whether the Application just has been created or existed for a while.

Reviewer 3: ```suggestion
Please note that any delay you configure here will delay _any_ automatic sync, regardless of whether the Application has just been created or has existed for a while.
```

### Block on empty dependencies

> It is possible to block syncing when dependency selectors are configured, but no dependency exists yet. This may be useful in scenarios where you want to prevent an Application to be synced unless another specific Application exists.

Reviewer 3: ```suggestion
It is possible to block syncing when dependency selectors are configured, but no dependency exists yet. This may be useful in scenarios where you want to prevent an Application from being synced unless another specific Application exists.
```

To configure blocked syncs on empty dependencies, use the field `blockOnEmpty` as follows:

```yaml
spec:
  dependsOn:
    blockOnEmpty: true
```

> Note that when you set `blockOnEmpty` to true, and sync this app through the CLI, the `argocd app sync` command will only return when the application starts syncing, i.e. when it has finished waiting. If you need the CLI to return immediately, use `argocd app sync --async`.

Reviewer 3: It feels like you want:
```suggestion
Note that when you set `blockOnEmpty` to true, and sync this app through the CLI, the `argocd app sync` command will only return when the application starts syncing, i.e. when it has finished waiting or times out. If you need the CLI to return immediately, use `argocd app sync --async`.
```

... see below

### Configure a wait timeout

> You can configure the maximum time that Argo CD should wait for dependencies to be resolved and the sync to start. The timeout also applies to the wait spent in blocking for empty dependencies.

Reviewer 3: would this change the behavior of `argocd app sync` (mentioned above)?

Author: Good question actually, I haven't tested behaviour with the CLI yet. I'll follow up on this.

# Application dependencies

> Argo CD lets you natively model dependencies between Applications. This mechanism is intended to replace the prior way of modeling dependencies when using the app-of-apps pattern with sync waves on the Application custom resources.

Reviewer 8: I think we probably need a full section describing the pros/cons of different ordering mechanisms: appset progressive syncs, app-of-apps w/ sync waves, app dependencies, and external tooling like CI / Kargo / Telefonistka. I'm not yet sure that app-of-apps w/ sync waves should be considered an anti-pattern. If it is, we should have a docs section explaining why.

Using the new dependency mechanism, users are not bound to the app-of-apps pattern anymore and can define dependencies between applications using application selectors to define a directed, acyclic graph for dependency modeling. This means that an Application which defines one or more dependencies do not only depend on its direct dependencies, but to the dependencies of its dependencies as well.

Application dependencies support both, manual and automatic sync policies.

## Sync behaviour

> Defining dependencies for an Application will affect its sync behaviour. When a sync is started, Argo CD will evaluate the sync and health status of all other Applications in its dependency tree. Only if all of its dependent Applications have a sync status of `Synced` and a health status of `Healthy`, will the sync proceed. Until all conditions are met, the sync will stay in a progressing state and a message will indicate that the sync is waiting for dependencies. This behaviour is similar to sync waves, where a sync progresses only to the next wave when the current wave's resources are all synced and healthy.

Reviewer 8: Does this mean that churn in dependencies could perpetually block an app from syncing?

For example, suppose we have apps A, B, and C in a simple chain. If A and B continually switch off with one syncing for a while, then the other, etc., will C ever sync? 

If not, we should probably document that edge case as potentially problematic.

> As a fallback mechanism, when using manual sync with the force option, Argo CD will not perform dependency resolution and the sync will proceed regardless of the states of the Application's dependencies.

Reviewer 8: I'm not sure we should reuse the force-sync option for this. Force sync has its own use cases which don't necessarily align with the use cases for skipping dependency evaluation. Instead, I think we should introduce a new sync flag called something like "skip dependencies."

## Feature state and caveats

> This feature is in [beta state](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md#beta). It will be polished across the upcoming releases.

Reviewer 8: I think this feature should start in alpha behind a feature flag. If the feature already included a way to handle cross-project dependencies, provided a full UI/CLI experience, and handled things like warning on deletion of a dependency app, I think it would make sense to start in beta.

Bug reports and enhancement requests are most welcome.

## Limitations and known issues

A list of things that are good to know before getting started with Application dependencies:

> * An Application can only depend on other Applications that use the same AppProject and are in the same namespace as the dependent Application.

Reviewer 8: I think we need a section to emphasize that this is _not_ a reason to start putting a bunch of apps into the same project when that would go against security best practices. We should stress that, if your apps don't make sense to be in the same project, this feature simply isn't for you.

> * A circular dependency defined anywhere in the dependency chain is considered to be an error and will prevent all applications in the circular dependency loop from syncing.

Reviewer 8: I think we need a DAG visualization tool, because otherwise resolving these circular dependencies could be very, very difficult.

!!!warning
    Using force sync to override application dependencies should only be used as a last resort, as it may break your applications in multiple ways.

## Configuring Application dependencies

The dependencies of an Application are configured using the field `dependsOn` in the Application's spec. The field `dependsOn` is a complex data type, which lets you define one or more `selectors` to match a set of other Applications that this Application will be dependent upon.

Each selector can match Applications according to one or more properties. If you specify more than one matcher in a selector, each matcher has to match in order for an Application to be considered a dependency. For example, if you specify a `labelSelector` and a `namePattern`, both will have to match in order for an Application to be selected.

Additionally, the behaviour of dependency resolution can be configured on a per Application basis.

## Dependency selectors

> Currently, the available properties to match against are `labels` and the `name` of the Application. Each selector can yield zero or more applications.

Reviewer 8: What is the behavior if my selectors match zero applications? Am I considered blocked because the dependency isn't satisfied, or unblocked because there are no dependencies?

Reviewer 8: I see it defaults to "don't block"`. Should probably mention `blockOnEmpty` here and maybe link to that section of the doc.

### Matching on labels

It is possible and recommended to use a standard label selector to select the dependencies for your Application:

```yaml
spec:
  dependsOn:
    selectors:
    - labelSelector:
        matchLabels:
            name: some-app
```

> The above YAML will have the Application depend on any other Application that matches the given label selector. In this case, any application carrying a label `name` with the value of `some-app`.

Reviewer 8: I think using `name` for a label example is a bit confusing.

### Matching on name patterns

Dependencies can be matched on the name of Applications by specifying the `namePattern` selector. The `namePattern` selector takes a list of Application name patterns to match against the Applications' `.metadata.name` property. Specifying multiple patterns 

```yaml
spec:
  dependsOn:
    selectors:
    - namePattern:
      - some-app
      - other-app
```

The above YAML will have the Application depend on the Applications `some-app` and `other-app`. 

> The `namePattern` matcher supports shell-style glob matching, so the following example would match all Applications whose name is starting with `stage1-`:

Reviewer 8: Can we specify the glob engine used to evaluate these and link to its docs? We use multiple glob engines for different purposes, so being specific should be quite helpful to the user.

```yaml
spec:
  dependsOn:
    selectors:
    - namePattern:
      - stage1-*
```

Regular expressions are not supported.

## Configuring behaviour of dependency resolution

There are several options to configure the behaviour of dependency resolution. Please note that these settings are not inherited to dependencies, but must be configured for each dependent uniquely.

### Delayed start of automated sync

In certain scenarios, you may want to delay the automatic sync process when there are dependencies configured between applications that are created at the same time (i.e. synced through an app of apps pattern from Git). This may allow the synchronization of all applications to succeed before the dependency tree is resolved and the sync is actually started in the right order.

To configure the sync delay, specify the number of seconds to delay in the field `syncDelay` as follows:

```yaml
spec:
  dependsOn:

>     syncDelay: 5

Reviewer 8: Maybe rename the field `syncDelaySeconds`?

### Block on empty dependencies

It is possible to block syncing when dependency selectors are configured, but no dependency exists yet. This may be useful in scenarios where you want to prevent an Application from being synced unless another specific Application exists.

To configure blocked syncs on empty dependencies, use the field `blockOnEmpty` as follows:

```yaml
spec:
  dependsOn:
    blockOnEmpty: true
```

Note that when you set `blockOnEmpty` to true, and sync this app through the CLI, the `argocd app sync` command will only return when the application starts syncing, i.e. when it has finished waiting or times out. If you need the CLI to return immediately, use `argocd app sync --async`.

### Configure a wait timeout

You can configure the maximum time that Argo CD should wait for dependencies to be resolved and the sync to start. The timeout also applies to the wait spent in blocking for empty dependencies.

To configure a timeout, set the field `timeout` to the number of seconds Argo CD should wait at most.

```yaml
spec:
  dependsOn:

>     timeout: 10

Reviewer 8: Maybe `timeoutSeconds`?

Application dependencies support both, manual and automatic sync policies.

## Sync behaviour

> Defining dependencies for an Application will affect that Application's sync behaviour. When a sync is started, Argo CD will evaluate the sync and health status of all other Applications in the dependency tree. Only if all of the other Applications that the Application is dependent upon have a sync status of `Synced` and a health status of `Healthy`, the sync will proceed. Until all conditions are met, the sync will stay in a progressing state and a message will indicate that the sync is waiting for dependencies. This behaviour is similar to sync waves, where a sync progresses only to the next wave when the current wave's resources are all synced and healthy.

Reviewer 4: small suggestion:
`s/affect that Application's/affect the Applications's`

Reviewer 3: I'm on the fence on this.

Author: ```suggestion
Defining dependencies for an Application will affect its sync behaviour. When a sync is started, Argo CD will evaluate the sync and health status of all other Applications in the dependency tree. Only if all of the other Applications that the Application is dependent upon have a sync status of `Synced` and a health status of `Healthy`, the sync will proceed. Until all conditions are met, the sync will stay in a progressing state and a message will indicate that the sync is waiting for dependencies. This behaviour is similar to sync waves, where a sync progresses only to the next wave when the current wave's resources are all synced and healthy.
```

@leoluz @jsoref How about this one instead?

Reviewer 3: ```suggestion
Defining dependencies for an Application will affect its sync behaviour. When a sync is started, Argo CD will evaluate the sync and health status of all other Applications in its dependency tree. Only if all of its dependent Applications have a sync status of `Synced` and a health status of `Healthy`, will the sync proceed. Until all conditions are met, the sync will stay in a progressing state and a message will indicate that the sync is waiting for dependencies. This behaviour is similar to sync waves, where a sync progresses only to the next wave when the current wave's resources are all synced and healthy.
```
